src/replication.c

```diff
@@ -1715,46 +1705,65 @@ void readSyncBulkPayload(connection *conn) {
      *    such case we want just to read the RDB file in memory. */
-    serverLog(LL_NOTICE, "MASTER <-> REPLICA sync: Flushing old data");
 
     /* We need to stop any AOF rewriting child before flushing and parsing
      * the RDB, otherwise we'll create a copy-on-write disaster. */
     if (server.aof_state != AOF_OFF) stopAppendOnly();
 
-    /* When diskless RDB loading is used by replicas, it may be configured
-     * in order to save the current DB instead of throwing it away,
-     * so that we can restore it in case of failed transfer. */
     if (use_diskless_load &&
         server.repl_diskless_load == REPL_DISKLESS_LOAD_SWAPDB)
     {
-        /* Create a backup of server.db[] and initialize to empty
-         * dictionaries. */
-        diskless_load_backup = disklessLoadMakeBackup();
+        /* Initialize empty tempDb dictionaries. */
+        diskless_load_tempDb = disklessLoadInitTempDb();
+    }
+    else
+    {
+        // TODO: what's the effect of not calling this on REPL_DISKLESS_LOAD_SWAPDB now?
+
+        /* OLD COMMENT: We call to emptyDb even in case of REPL_DISKLESS_LOAD_SWAPDB
+        * (Where disklessLoadMakeBackup left server.db empty) because we
+        * want to execute all the auxiliary logic of emptyDb (Namely,
+        * fire module events) */
+
+        serverLog(LL_NOTICE, "MASTER <-> REPLICA sync: Flushing old data");
+
+        emptyDb(-1,empty_db_flags,replicationEmptyDbCallback);
     }
-    /* We call to emptyDb even in case of REPL_DISKLESS_LOAD_SWAPDB
-     * (Where disklessLoadMakeBackup left server.db empty) because we
-     * want to execute all the auxiliary logic of emptyDb (Namely,
-     * fire module events) */
-    emptyDb(-1,empty_db_flags,replicationEmptyDbCallback);
 
     /* Before loading the DB into memory we need to delete the readable
-     * handler, otherwise it will get called recursively since
-     * rdbLoad() will call the event loop to process events from time to
-     * time for non blocking loading. */
+    * handler, otherwise it will get called recursively since
+    * rdbLoad() will call the event loop to process events from time to
+    * time for non blocking loading. */
     connSetReadHandler(conn, NULL);
+    
     serverLog(LL_NOTICE, "MASTER <-> REPLICA sync: Loading DB in memory");
     rdbSaveInfo rsi = RDB_SAVE_INFO_INIT;
     if (use_diskless_load) {
         rio rdb;
+        redisDb *dbarray;
+        int asyncLoading = 0;
+        if (server.repl_diskless_load == REPL_DISKLESS_LOAD_SWAPDB)
+        {
+            /* Async loading means we continue serving read commands during full resync.
+            * Enabled only on SWAPDB diskless replication when master replication ID hasn't changed. */
+            if (strcmp(server.replid, server.master_replid) == 0) {
+                asyncLoading = 1;
+            }
+            dbarray = diskless_load_tempDb->dbarray;
+        }
+        else
+        {
```

Comment:
some styling issues.. i can also fix later, when we're near complete...
```suggestion
        } else {
```